File: internal/score/score.go
package score

import (
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"
	"unicode"

	"github.com/mmcdole/gofeed"
)

const (
	SchemaVersion  = 2
	UserAgent      = "GardenRSS/0.1 (+https://github.com/newtosh/gardenrss; feed-health)"
	DefaultTimeout = 15 * time.Second

	Window1d  = 24 * time.Hour
	Window7d  = 7 * 24 * time.Hour
	Window30d = 30 * 24 * time.Hour
	Window90d = 90 * 24 * time.Hour
)

type HealthStatus string

const (
	HealthOK        HealthStatus = "ok"
	HealthStale     HealthStatus = "stale"
	HealthUnhealthy HealthStatus = "unhealthy"
)

type ReasonCode string

const (
	ReasonOK          ReasonCode = "ok"
	ReasonStale       ReasonCode = "stale"
	ReasonTimeout     ReasonCode = "timeout"
	ReasonDNS         ReasonCode = "dns"
	ReasonTLS         ReasonCode = "tls"
	ReasonHTTPStatus  ReasonCode = "http_status"
	ReasonTooLarge    ReasonCode = "too_large"
	ReasonBlockedURL  ReasonCode = "blocked_url"
	ReasonUnparseable ReasonCode = "unparseable"
	ReasonFetch       ReasonCode = "fetch_error"
)

// Result is the versioned per-feed score record (schemaVersion 2).
type Result struct {
	SchemaVersion   int          `json:"schemaVersion"`
	XMLURL          string       `json:"xmlUrl"`
	Health          HealthStatus `json:"health"`
	Reason          ReasonCode   `json:"reason"`
	VelocityUnknown bool         `json:"velocityUnknown"`
	Posts1d         *int         `json:"posts1d,omitempty"`
	Posts7d         *int         `json:"posts7d,omitempty"`
	Posts30d        *int         `json:"posts30d,omitempty"`
	AvgWords        *float64     `json:"avgWords,omitempty"`
	LastDatedAt     *time.Time   `json:"lastDatedAt,omitempty"`
	Title           string       `json:"title,omitempty"`
	Detail          string       `json:"detail,omitempty"`
	ScoredAt        time.Time    `json:"scoredAt"`
}

var tagStripper = regexp.MustCompile(`(?s)<[^>]*>`)

// ScoreParsedFeed computes health + multi-window cadence + word burden.
func ScoreParsedFeed(xmlURL string, feed *gofeed.Feed, now time.Time) Result {
	r := Result{
		SchemaVersion: SchemaVersion,
		XMLURL:        xmlURL,
		Health:        HealthOK,
		Reason:        ReasonOK,
		ScoredAt:      now.UTC(),
	}
	if feed == nil {
		r.Health = HealthUnhealthy
		r.Reason = ReasonUnparseable
		r.VelocityUnknown = true
		return r
	}
	if feed.Title != "" {
		r.Title = feed.Title
	}

	start1 := now.Add(-Window1d)
	start7 := now.Add(-Window7d)
	start30 := now.Add(-Window30d)
	start90 := now.Add(-Window90d)

	var posts1, posts7, posts30, posts90 int
	var anyDated bool
	var lastDated *time.Time
	var wordSum int
	var wordItems int

	for _, item := range feed.Items {
		t := itemPublished(item)
		if t == nil {
			continue
		}
		anyDated = true
		if lastDated == nil || t.After(*lastDated) {
			cp := *t
			lastDated = &cp
		}
		if !t.Before(start1) && !t.After(now) {
			posts1++
		}
		if !t.Before(start7) && !t.After(now) {
			posts7++
		}
		if !t.Before(start30) && !t.After(now) {
			posts30++
		}
		if !t.Before(start90) && !t.After(now) {
			posts90++
		}
		if n := wordCount(bestItemText(item)); n > 0 {
			wordSum += n
			wordItems++
		}
	}

	if !anyDated {
		// No publish dates (even after permalink inference): treat as stale
		// so OPML owners get Fix URL instead of silent cadence-unknown healthy.
		r.Health = HealthStale
		r.Reason = ReasonStale
		r.VelocityUnknown = true
		return r
	}

	r.Posts1d = &posts1
	r.Posts7d = &posts7
	r.Posts30d = &posts30
	r.VelocityUnknown = false
	if lastDated != nil {
		utc := lastDated.UTC()
		r.LastDatedAt = &utc
	}
	if wordItems > 0 {
		avg := float64(wordSum) / float64(wordItems)
		avg = float64(int(avg*100+0.5)) / 100
		r.AvgWords = &avg
	}

	if posts90 == 0 {
		r.Health = HealthStale
		r.Reason = ReasonStale
	}
	return r
}

func Unhealthy(xmlURL string, reason ReasonCode, now time.Time) Result {
	return Result{
		SchemaVersion:   SchemaVersion,
		XMLURL:          xmlURL,
		Health:          HealthUnhealthy,
		Reason:          reason,
		VelocityUnknown: true,
		ScoredAt:        now.UTC(),
	}
}

func itemPublished(item *gofeed.Item) *time.Time {
	if item == nil {
		return nil
	}
	if item.PublishedParsed != nil {
		return item.PublishedParsed
	}
	if item.UpdatedParsed != nil {
		return item.UpdatedParsed
	}
	if t := inferDateFromPermalink(item.Link); t != nil {
		return t
	}
	return nil
}

// inferDateFromPermalink reads /YYYY/MM/DD/ or /YYYY-MM-DD/ from item links.
func inferDateFromPermalink(link string) *time.Time {
	if link == "" {
		return nil
	}
	path := link
	if u, err := url.Parse(link); err == nil {
		path = u.Path
	}
	reSlash := regexp.MustCompile(`/((?:19|20)\d{2})/([01]?\d)/([0-3]?\d)(?:/|$)`)
	reDash := regexp.MustCompile(`/((?:19|20)\d{2})-([01]?\d)-([0-3]?\d)(?:/|$)`)
	m := reSlash.FindStringSubmatch(path)
	if m == nil {
		m = reDash.FindStringSubmatch(path)
	}
	if m == nil {
		return nil
	}
	y, _ := strconv.Atoi(m[1])
	mo, _ := strconv.Atoi(m[2])
	d, _ := strconv.Atoi(m[3])
	if mo < 1 || mo > 12 || d < 1 || d > 31 {
		return nil
	}
	t := time.Date(y, time.Month(mo), d, 0, 0, 0, 0, time.UTC)
	if t.Year() != y || int(t.Month()) != mo || t.Day() != d {
		return nil
	}
	return &t
}

func bestItemText(item *gofeed.Item) string {
	if item == nil {
		return ""
	}
	if item.Content != "" {
		return item.Content
	}
	if item.Description != "" {
		return item.Description
	}
	return ""
}

func wordCount(raw string) int {
	plain := tagStripper.ReplaceAllString(raw, " ")
	plain = htmlEntityRough(plain)
	n := 0
	inWord := false
	for _, r := range plain {
		if unicode.IsSpace(r) {
			inWord = false
			continue
		}
		if !inWord {
			n++
			inWord = true
		}
	}
	return n
}

func htmlEntityRough(s string) string {
	s = strings.ReplaceAll(s, "&nbsp;", " ")
	s = strings.ReplaceAll(s, "&amp;", "&")
	s = strings.ReplaceAll(s, "&lt;", "<")
	s = strings.ReplaceAll(s, "&gt;", ">")
	s = strings.ReplaceAll(s, "&quot;", "\"")
	return s
}

File: workers/score/src/score.ts
import type { ParsedFeed, ParsedItem, ReasonCode, ScoreResult } from './types'
import {
  SCHEMA_VERSION,
  WINDOW_1D_MS,
  WINDOW_7D_MS,
  WINDOW_30D_MS,
  WINDOW_90D_MS,
} from './types'

/** Score a successfully parsed feed (parity with Go ScoreParsedFeed). */
export function scoreParsedFeed(
  xmlUrl: string,
  feed: ParsedFeed | null,
  now: Date = new Date(),
): ScoreResult {
  const scoredAt = now.toISOString().replace(/\.\d{3}Z$/, 'Z')
  if (feed == null) {
    return unhealthy(xmlUrl, 'unparseable', now)
  }

  const result: ScoreResult = {
    schemaVersion: SCHEMA_VERSION,
    xmlUrl,
    health: 'ok',
    reason: 'ok',
    velocityUnknown: false,
    scoredAt,
  }
  if (feed.title) {
    result.title = feed.title
  }

  const start1 = new Date(now.getTime() - WINDOW_1D_MS)
  const start7 = new Date(now.getTime() - WINDOW_7D_MS)
  const start30 = new Date(now.getTime() - WINDOW_30D_MS)
  const start90 = new Date(now.getTime() - WINDOW_90D_MS)

  let posts1 = 0
  let posts7 = 0
  let posts30 = 0
  let posts90 = 0
  let anyDated = false
  let lastDated: Date | null = null
  let wordSum = 0
  let wordItems = 0

  for (const item of feed.items) {
    const t = item.published
    if (t == null) continue
    anyDated = true
    if (lastDated == null || t > lastDated) lastDated = t
    if (t >= start1 && t <= now) posts1++
    if (t >= start7 && t <= now) posts7++
    if (t >= start30 && t <= now) posts30++
    if (t >= start90 && t <= now) posts90++
    const n = wordCount(bestItemText(item))
    if (n > 0) {
      wordSum += n
      wordItems++
    }
  }

  if (!anyDated) {
    // No publish dates (even after permalink inference): treat as stale so
    // owners get Fix URL instead of a silent "Cadence unknown" healthy row.
    result.health = 'stale'
    result.reason = 'stale'
    result.velocityUnknown = true
    result.detail = 'No publish dates found in feed'
    return result
  }

  result.posts1d = posts1
  result.posts7d = posts7
  result.posts30d = posts30
  result.velocityUnknown = false
  if (lastDated) {
    result.lastDatedAt = lastDated.toISOString().replace(/\.\d{3}Z$/, 'Z')
  }
  if (wordItems > 0) {
    result.avgWords = Math.round((wordSum / wordItems) * 100) / 100
  }
  if (posts90 === 0) {
    result.health = 'stale'
    result.reason = 'stale'
  }
  return result
}

export function unhealthy(
  xmlUrl: string,
  reason: ReasonCode,
  now: Date = new Date(),
  detail?: string,
): ScoreResult {
  return {
    schemaVersion: SCHEMA_VERSION,
    xmlUrl,
    health: 'unhealthy',
    reason,
    velocityUnknown: true,
    scoredAt: now.toISOString().replace(/\.\d{3}Z$/, 'Z'),
    ...(detail ? { detail } : {}),
  }
}

function bestItemText(item: ParsedItem): string {
  if (item.content) return item.content
  if (item.description) return item.description
  return ''
}

function wordCount(raw: string): number {
  const plain = raw
    .replace(/<[^>]*>/gs, ' ')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
  const parts = plain.trim().split(/\s+/).filter(Boolean)
  return parts.length
}

export type { ParsedFeed, ParsedItem, ScoreResult }
